### PR TITLE
Update QM driver for new channel API

### PR DESCRIPTION
**Pull request by stavros11**
I would still like to try cleaning up a few places in terms of code, so reviews can wait, however the primary goal which was to run something on hardware is achieved, at least for the single shot classification experiment. Sweepers are still to be tested and we cannot run qibocal routines because they need to be updated for the new pulse and channel API. I may try to update some to check how everything works.

Testing platform: qw11qD from https://github.com/qiboteam/qibolab_platforms_qrc/pull/153
<details>
<summary>Testing script for single shot experiment (written in the 0.2 API)</summary>

```py
import numpy as np
from qibolab import create_platform, ExecutionParameters
from qibolab.pulses import PulseSequence, Delay

platform = create_platform("qw11qD")
working_qubits = ["D1", "D2", "D3", "D4"]

ro_pulses = {} # save ids of readout pulses so that we can grab results
sequence0 = PulseSequence()
sequence1 = PulseSequence()
for q in working_qubits:
    qubit = platform.qubits[q]

    sequence0.extend(qubit.native_gates.MZ.create_sequence())

    rseq = qubit.native_gates.RX.create_sequence(theta=np.pi, phi=0)
    mseq = qubit.native_gates.MZ.create_sequence()
    sequence1.extend(rseq)
    sequence1[qubit.measure.name].append(Delay(duration=rseq.duration))
    sequence1.extend(mseq)

    ro_pulses[q] = mseq[qubit.measure.name][0].id

platform.connect()

options = ExecutionParameters(nshots=5000, relaxation_time=100000)
result0 = platform.sweep(sequence0, options)
result1 = platform.sweep(sequence1, options)

platform.disconnect()

for q in working_qubits:
    output0 = result0[ro_pulses[q]]
    output1 = result1[ro_pulses[q]]
    print(q)
    print(np.unique(output0.samples, return_counts=True))
    print(np.unique(output1.samples, return_counts=True))
    print()
```

Expected output (confirmed):
```
...
Connected to: Rohde&Schwarz SGS100A (serial:1416.0505k02/114164, firmware:4.2.76.0-4.30.046.295) in 0.19s
2024-07-13 19:47:37,351 - qm - INFO     - Sending program to QOP for compilation
2024-07-13 19:47:37,592 - qm - INFO     - Executing program
2024-07-13 19:47:40,503 - qm - INFO     - Sending program to QOP for compilation
2024-07-13 19:47:40,952 - qm - INFO     - Executing program
D1
(array([0, 1], dtype=uint32), array([4901,   99]))
(array([0, 1], dtype=uint32), array([ 247, 4753]))

D2
(array([0, 1], dtype=uint32), array([4916,   84]))
(array([0, 1], dtype=uint32), array([ 248, 4752]))

D3
(array([0, 1], dtype=uint32), array([4832,  168]))
(array([0, 1], dtype=uint32), array([ 274, 4726]))

D4
(array([0, 1], dtype=uint32), array([4971,   29]))
(array([0, 1], dtype=uint32), array([ 886, 4114]))
```

</details>

Features supported in 0.1 that are not yet supported here:
* Unrolling*,
* Baking,
* Helper functions for Octave mixer calibration (need to rewrite because these weren't very useful in 0.1 either),
* Using OPX+ with external mixers and LOs (instead of Octaves) for drive and measurement.

*Unrolling may work if a single unrolled sequence is passed, but has not been tested. I noticed that #885 changed the API and now the `Platform` is passing a list of sequences to the instrument. I tend to agree with this for reasons not directly related to 0.2, however for now does it mean that we need to unroll the list in every driver?

Tests are failing on the base branch, so I will wait until they pass there to update. Most QM specific tests need to be updated as well, but I'd like to come up with a good strategy for testing without connecting to instruments before refactoring them.

### CODE REVIEW

**Comment by stavros11** [comment]
Some working qibocal routines (on qibocal branch: `0.2-channels`):
* Resonator spectroscopy: http://login.qrccluster.com:9000/a2yqjfi-QVeOwzg5343cbg==
* Rabi amplitude signal: http://login.qrccluster.com:9000/TPCAj9DPTra19La2AcrlMA==
* Single shot classification: http://login.qrccluster.com:9000/dtP-ggnrRo6RDTq4wcUxZQ==
* Rabi amplitude probabilities: http://login.qrccluster.com:9000/00doS90-TE2SOrKK9AJtog==

**Comment by alecandido** [comment]
> *Unrolling may work if a single unrolled sequence is passed, but has not been tested. I noticed that #885 changed the API and now the `Platform` is passing a list of sequences to the instrument. I tend to agree with this for reasons not directly related to 0.2, however for now does it mean that we need to unroll the list in every driver?

Not sure about which are your reasons (and I'd be interested in hearing them), but unrolling in every driver seems a great loss, I'd be quite against it...

**Comment by stavros11** [comment]
> Not sure about which are your reasons (and I'd be interested in hearing them), 

If I ever need to address performance issues such as #912 at the driver level, it would be more convenient to have access to the list of sequences instead of a single unrolled sequence. However, I am not planning to do this in the near term and I am not sure if it will be done for 0.2.x either.

> but unrolling in every driver seems a great loss, I'd be quite against it...

I agree with this. But then what is the point of passing a list of sequences in https://github.com/qiboteam/qibolab/blob/663f790c7964ce491a1015c177e8e1a41a8b9766/src/qibolab/platform/platform.py#L361-L367
?

That being said, with the channels almost finalized, now it might be a good time to fix the execution interface (complete #861), so that we only have to implement a single execution method in the drivers. I think that as it is now `platform.execute_pulse_sequence` is broken, so I had to redirect it to `platform.sweep` in 443dac8c8fa5dc5b97867fcc856df1943c78b37c to run qibocal routines. But we shouldn't need to maintain all different methods in the platform and instruments if we are eventually going to drop them.

**Comment by alecandido** [comment]
> If I ever need to address performance issues such as #912 at the driver level, it would be more convenient to have access to the list of sequences instead of a single unrolled sequence.

I'm not sure that, at this stage, there would be a significant gain in having the individual sequences at hand.
For sure, we should reconsider the unrolling with #917, but this is truly Qibolab 0.3, so it is not something we should worry about now.

> That being said, with the channels almost finalized, now it might be a good time to fix the execution interface (complete #861), so that we only have to implement a single execution method in the drivers.

It was my intention to close that, and implement the interface for #692 immediately (not waiting for channels). At least, they will be two things less on the list, and you can already start adapting the QM driver for those (though the unique execution interface is mostly not affecting the drivers).

**Comment by stavros11** [comment]
I have rebased this PR to 0.2 after the latest merges and retried some of the qibocal routines I have tested before (required updating https://github.com/qiboteam/qibocal/pull/809 as well). The following worked:
* resonator spectroscopy: https://github.com/qiboteam/qibocal/pull/809 (errors are wrong because I disabled their calculation, as I am not sure how std was calculated before with cyclic averaging mode. If errors are disabled in the plot the acquisition looks correct).
* rabi signal: http://login.qrccluster.com:9000/789GAdg_Ssm8ukc2XO8s4Q==
* single shot: http://login.qrccluster.com:9000/U5lzU6UwSDS2jaoC8FiLCw==
* rabi probabilities: http://login.qrccluster.com:9000/9NqMuDQ4QoGriZA7XCml7g==

I will now mark this as ready for review as it is already a big change and it is important for other developments.

Here is a list of issues that I am planning to implement in other PRs in that order (will convert to issues if this is merged before addressing them):
- [ ] Pulse scheduling with duration sweepers (#935),
- [ ] Baking: support for durations that are not multiple of 4ns (not sure to what extend yet),
- [ ] Parallel sweeps: the sweeper structure in the driver is updated with this in mind, however I did not fully implement it for the case `len(parallel_sweepers) > 1` mainly because there is no qibocal routine using this.

**Comment by stavros11** [comment]
Regarding the new structure of the driver, the main idea is that:
* `components/` implements the QM specialization of the `Channel` and some special configs (as per the new channel structure),
* `config/` contains the `dataclass`es required to generate the QM configuration dictionary (which is generated using `asdict`),
* `program/` contains various functions that are generating the QUA program (with sweeps if needed),
* `controller.py` implements the main `Controller` instrument, the `Octave` class for registering Octaves in the platform.

The `controller` has access to everything. There is no overlap between the `program` and `config`, apart from the following tiny function https://github.com/qiboteam/qibolab/blob/130ac056e6ceccf0e2bd842fa7549d141163ac5e/src/qibolab/instruments/qm/config/pulses.py#L21-L23
which is used in both.

**Comment by alecandido** [comment]
Luckily, there were truly few conflicts with #956.

In case you want to make sure everything was smooth, I saved the [diff](https://github.com/user-attachments/files/16530186/pr956.diff.zip) with the previous state of this PR, that should correspond to just the content introduced in #956 (sorry, colors and rich text are gone, so it's a less pleasant reading...).

I will now finish the review of this as well. Hopefully it was the last rebase.

(in any case, despite being non-trivial nor small, this PR is touching a very isolated portion of Qibolab, so conflicts are few in general, and the rebase is mostly automatic)

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [inline src/qibolab/instruments/qm/config/config.py:90]: What about splitting this to a custom constructor for `AcquireOctaveElement`?

(same above for `RfOctaveElement`, but it's much smaller)

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [inline src/qibolab/instruments/qm/config/config.py:98]: I'd split this as a separate operation. Conceptually it is.

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [inline src/qibolab/instruments/qm/config/config.py:106]: Same as above. But you may need just a flag to exclude Q, and reuse the same operation.

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [inline src/qibolab/instruments/qm/config/config.py:120]: Same as above, with a toggle to switch from `QmPulse` to `QmAcquisition`.

File: src/qibolab/instruments/qm/program/acquisition.py
Comment (alecandido) [inline src/qibolab/instruments/qm/program/acquisition.py:29]:
```suggestion
        return data[..., :npulses, :]
```
?

This isn't the exact same, since it's transposed, and the result is an array, and not a list.

The exact equivalent is:
```suggestion
        return list(np.moveaxis(data, -2, 0)[:npulses])
```
but, if the dimension contains exactly `npulses` elements, you could even skip the further `[:npulses]`.

Comment (alecandido) [reply src/qibolab/instruments/qm/program/acquisition.py:29]: Initially, I wrote a more complicate expression with `np.transpose()`, but I was sure there should have been an easier way.

And now that I found `np.moveaxis()`, I should also use it for:
https://github.com/qiboteam/qibolab/blob/1e3a501aa43f907a2415f1ebe903c3853a093a38/src/qibolab/result.py#L16
:DD

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (alecandido) [inline src/qibolab/instruments/qm/program/sweepers.py:87]: What about replacing this class with plain functions?

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (alecandido) [inline src/qibolab/instruments/qm/program/sweepers.py:73]: For this, you just need a mapping: `Sweeper.parameter -> qua_variable_type`.

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (alecandido) [inline src/qibolab/instruments/qm/program/sweepers.py:87]: This could be just a function, one for each parameter type.

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (alecandido) [inline src/qibolab/instruments/qm/program/sweepers.py:81]: And this is just a function to postprocess values. Which in most cases could be just the identity `lambda x: x`.

**Review by alecandido — COMMENTED** [review]
I had a chance to have another look at sweepers.

Classes are not bad in general, and this case is borderline, since you'd still need a map and a few functions around.
However, the proposed changes are not improving that much the code. But taking a simplification into account might suggest other features that are possible to be lifted to the general Qibolab, reducing even further the gap, and providing comfortable helpers even to the other drivers.

File: src/qibolab/instruments/qm/config/pulses.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/pulses.py]: 64bd4d6a58ca729c8fd67976071b302d2637c881

File: src/qibolab/instruments/qm/components/channel.py
Comment (stavros11) [reply src/qibolab/instruments/qm/components/channel.py]: 3285283d4b33db8eca800047d2c3ccb640732427

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: 4fe1ffdb501b223da3babf38ffdb25bcc26a1920

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: 4fe1ffdb501b223da3babf38ffdb25bcc26a1920

File: src/qibolab/instruments/qm/config/pulses.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/pulses.py]: d5ef6305bd9d4c381d1d98bc426f07de913d058a

File: src/qibolab/instruments/qm/config/pulses.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/pulses.py]: d5ef6305bd9d4c381d1d98bc426f07de913d058a

File: src/qibolab/instruments/qm/config/devices.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/devices.py]: 4fe1ffdb501b223da3babf38ffdb25bcc26a1920

File: src/qibolab/instruments/qm/config/devices.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/devices.py]: If the Octave is used for readout, they need to be overwritten in order to set the gain, if this was specified by the user in the platform definition.

File: src/qibolab/instruments/qm/config/devices.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/devices.py]: I believe there is no port 0 for QM. Output ports are 1-10 for OPX+ and 1-5 for Octaves. If someone uses `port = 0` here, it will result to a `"0"` and a `"-1"` somewhere in the config which I hope (but have not tested) that it will result to an error from the qm-qua side, since they are validating the config. 

Otherwise, we should raise the errors ourselves, but as long as they are doing it, I would save us the additional lines of code.

Comment (stavros11) [reply src/qibolab/instruments/qm/config/devices.py]: Also the hardcoded `2 * port - 1`, `2 * port` is due to the way OPX+ ports are connected to Octaves. In principle, you can do screw this up changing these cables, since they are external (not inside the instruments). However, I believe this kind of connectivity is the one suggested by QM and also the most sensible from cable organization point of view, so I decided to also hardcode it here, instead of complicating the driver more to support something that is highly unlikely to be useful.

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: ee8087b67ed88e2f1e003560c0e6449e10bc7d69

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: 940dfc78fdaa65b0e52d651bc37db955872903a6

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: 940dfc78fdaa65b0e52d651bc37db955872903a6

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: 940dfc78fdaa65b0e52d651bc37db955872903a6

File: src/qibolab/instruments/qm/config/devices.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/devices.py:23]: Could be a bit misleading, because in principle here we are not directly working with JSONs. It is just that qm-qua complains if I pass a `config` with integers as keys, they need strings everywhere, even for port identifiers. I am not sure if this is because they are internally converting to JSON (very likely) or some other reason.

File: src/qibolab/instruments/qm/program/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/program/acquisition.py]: ad5cfee753b2968740c42c919c8c8cad2d8fa70b

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (stavros11) [reply src/qibolab/instruments/qm/program/sweepers.py]: f6d3b5c16d366e2bef082d90a9ea4956a255b5ab

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (stavros11) [reply src/qibolab/instruments/qm/program/sweepers.py]: f6d3b5c16d366e2bef082d90a9ea4956a255b5ab

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (stavros11) [reply src/qibolab/instruments/qm/program/sweepers.py]: f6d3b5c16d366e2bef082d90a9ea4956a255b5ab

File: src/qibolab/instruments/qm/program/sweepers.py
Comment (stavros11) [reply src/qibolab/instruments/qm/program/sweepers.py]: f6d3b5c16d366e2bef082d90a9ea4956a255b5ab

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py:30]: As far as I recall, this is a peculiarity of QM and needs to be registered for things to work. I believe during the initial installation they suggested to use it without much explanation why (or I don't remember the explation). I am not sure if this is still the case for the latest qm-qua versions, however I believe we now need this anyway because we are using digital triggers to allow the Octave's LO signal to pass when needed.

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py:37]: > `Controller` is not frozen, and it is plenty of mutable elements (dicts and strings).
> 
> Maybe, if you're not overwriting it, you shouldn't do it silently. At least, return a boolean to tell whether it was already present.
> 
> Btw, which is the reason to avoid overwriting?

With the current design we are registering everything by looping over the sequence and registering only devices and channels that are being used, except DC channels with offsets that are always registerd in order to keep qubits at their sweetspots even when they are not used in the sequence. If we overwrite the controller entry here, we will potentially lose ports that were registered from a previous pulse on the same controller.

Regarding the boolean, I could easily return it, but I think that the other method that is calling this would probably just ignore it. If that's the case, what would be the use of returning?

File: src/qibolab/instruments/qm/config/devices.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/devices.py:23]: Fine both ways.
You could even call `StrKeyDict`. But we could wait for when this will be used beyond the ports.

*I'm pretty sure it's JSON, also because the docs seem based on OpenAPI (and in general it makes sense). However, it's pretty much irrelevant for us*

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: > Here I'm just puzzled, but I should think about it more carefully (so, the comment is also a reminder to myself).
> 
> It's true that you're configuring an IQ line, and that's a modular operation. However, in order to configure the line, you need the octave to be already partially registered. But it's configuration as an LO is coming only here.
> 
> I'd have expected that either the octave had to be fully registered in advance, or it would be fully configured here (which doesn't seem to be possible, since the connectivity might involve multiple lines).

The Octave is registered in advance calling `config.add_octave`. The output ports are registered by this function which I believe has all the required information (LO configuration, IF, etc.).

> Also, what it happens if the octave is configured on multiple lines, with incompatible OscillatorConfigs?
I guess not all the lines are distinguished by (dev, port), since that should be different only for physically different lines. But maybe I'm just wrong...

Here you are partly right. The idea is that the user is aware of how many LOs are available in the instrument and creates the components appropriately to avoid collisions: https://github.com/qiboteam/qibolab_platforms_qrc/blob/0.2-qw11qD/qw11qD/parameters.json. I guess if this is not done properly, they may end up overwriting some LO configuration by mistake. For Octaves in particular, this is a bit trickier because as discussed before ports 2, 3 and 4, 5 share LOs even though we are allowed to pass different LO frequencies in the config (one of them is ignored). In principle, this is a problem from the qm-qua side and I hope they will raise appropriate errors in later releases (that's not the case so far). Otherwise the only other solution is to raise the errors ourselves.

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py:15]: Since this is fully internal, let me postpone this and the other pydantic related comment to a later release: https://github.com/qiboteam/qibolab/issues/967.

File: src/qibolab/instruments/qm/config/elements.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/elements.py:60]: https://github.com/qiboteam/qibolab/issues/967

File: src/qibolab/instruments/qm/config/pulses.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/pulses.py:58]: I guess we can easily lift this to the general pulse library. On the other hand, we could also wait to see if this is actually useful for other instruments before doing so.

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/config.py:30]: Ok, that's a bit weird, but if that's the case, there is not much we can do.

The only possible improvement would be to move the dictionary as a top-level constant, and document. But the only sensible piece of information to add in these docs is the part about the Octave, and the claim "you may always need this in any case".

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/config.py:37]: > With the current design we are registering everything by looping over the sequence and registering only devices and channels that are being used, except DC channels with offsets that are always registerd in order to keep qubits at their sweetspots even when they are not used in the sequence. If we overwrite the controller entry here, we will potentially lose ports that were registered from a previous pulse on the same controller.

Ok, but instead of making unneeded calls to `.add_x()`, and ignore them, I'd prefer to deduplicate them ahead of time. It's less surprising.

> Regarding the boolean, I could easily return it, but I think that the other method that is calling this would probably just ignore it. If that's the case, what would be the use of returning?

In principle, because you don't know who your caller is. Otherwise, it's not a modular design, and you reduce part of the advantage of having a separate function.

File: src/qibolab/instruments/qm/config/pulses.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/pulses.py:58]: > On the other hand, we could also wait to see if this is actually useful for other instruments before doing so.

I partially agree with you. Though, this is *potentially* useful, and the price of lifting is almost negligible (in terms of code and readability). Thus, in general, I'd work in advance to prevent duplication.

In any case, it's truly not that relevant.

File: src/qibolab/instruments/qm/config/devices.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/devices.py]: Ok, here I accept your explanation. In any case, if someone with a QM cluster would like to use Qibolab, but they cabled the connections differently, they should be capable enough to get in touch with us (and it wouldn't take that much to fix it).

File: src/qibolab/instruments/qm/config/devices.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/devices.py]: Ok, fine :)

File: src/qibolab/instruments/qm/config/config.py
Comment (alecandido) [reply src/qibolab/instruments/qm/config/config.py]: Ok, I see it's not trivial at all, and there is no urgent need to change the current status.

Thanks for the explanation, I'll just keep that in mind!

**Review by alecandido — APPROVED** [review]
Thanks for the PR in general, this is making the QM driver even more polished, and that makes it easier for other people to contribute. It's definitely a valuable effort!

Most of the replies left are of the kind "we may be able to do something, but it's not too relevant".

Merge whenever you wish :)

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py]: 4df6925e98093831c7242c27fcd44087c9ac6cd1

File: src/qibolab/instruments/qm/config/config.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/config.py:44]: > Ok, but instead of making unneeded calls to `.add_x()`, and ignore them, I'd prefer to deduplicate them ahead of time. It's less surprising.

I would agree, particularly with this point. I was planning to implement directly but it is probably slightly more involved than what I initially thought so I moved to later milestone (#968) for now, as it is fully internal. It is not very complicated though, so I may end up doing earlier.

File: src/qibolab/instruments/qm/config/pulses.py
Comment (stavros11) [reply src/qibolab/instruments/qm/config/pulses.py]: Implemented in 22d599307b44d689774b262d13e8ade07f2cbc37 and tested that it doesn't break anything. I will now merge this PR, if you like to change something in that implementation, we can do so in a different PR.

**Pull request merged**
